### PR TITLE
feat: Add LSP for FTL

### PR DESCRIPTION
**Pull request by worstell**
Add an LSP that produces document highlights when FTL outputs build errors

this is a work in progress

### CODE REVIEW

File: lsp/lsp.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	"github.com/TBD54566975/ftl/go-runtime/compile"
+	ftlErrors "github.com/TBD54566975/ftl/internal/errors"
+)
+
+const lsName = "ftl-language-server"
+
+// Server is a language server.
+type Server struct {
+	Server      *glspServer.Server
+	GlspLogger  *GLSPLogger
```
Comment (alecthomas) [inline lsp/lsp.go:23]: Should these be private?

File: lsp/lsp.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+		if errors.As(subErr, &ce) {
+			cp := ce.Pos
+			if errByFilename[cp.Filename] == nil {
+				errByFilename[cp.Filename] = errSet{}
+			}
+			errByFilename[cp.Filename][strings.TrimSpace(ce.Error())] = ce
+		}
+	}
+
+	go func() {
```
Comment (alecthomas) [inline lsp/lsp.go:68]: Maybe pull the body of this out into a separate function.

File: lsp/lsp.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+		for filename, errs := range errByFilename {
+			var diagnostics []protocol.Diagnostic
+			for _, e := range errs {
+				pp := e.Pos
+				sourceName := "ftl"
+				severity := protocol.DiagnosticSeverityError
+				diagnostics = append(diagnostics, protocol.Diagnostic{
+					Range: protocol.Range{
+						Start: protocol.Position{Line: protocol.UInteger(pp.Line), Character: protocol.UInteger(pp.Column + pp.Offset)},
+						End:   protocol.Position{Line: protocol.UInteger(pp.Line), Character: protocol.UInteger(pp.Column + pp.Offset + 10)}, //todo: fix
```
Comment (alecthomas) [inline lsp/lsp.go:78]: We should introduce a `EndCol` field into the positional information at some point.

File: lsp/lsp.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	return server
+}
+
+func (s *Server) Run() error {
+	return errors.Wrap(s.Server.RunStdio(), "lsp")
+}
+
+type errSet map[string]compile.Error
+
+func (s *Server) Post(err error) {
```
Comment (alecthomas) [inline lsp/lsp.go:53]: Add a comment describing what `err` needs to be, ie. joined `compile.Error`s

File: lsp/lsp.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	// handler.TextDocumentDidOpen = server.textDocumentDidOpen()
+	handler.TextDocumentDidChange = server.textDocumentDidChange()
+	return server
+}
+
+func (s *Server) Run() error {
+	return errors.Wrap(s.Server.RunStdio(), "lsp")
+}
+
+type errSet map[string]compile.Error
```
Comment (alecthomas) [inline lsp/lsp.go:51]: Factor `compile.Error` out into a common package - I might put it in `schema`

**Review by alecthomas — APPROVED** [review]

File: go-runtime/compile/schema.go [hunk 6]
```diff
@@ -621,13 +629,13 @@ func visitStruct(pctx *parseContext, pos token.Pos, tnode types.Type) (*schema.R
...
 			metadata = append(metadata, &schema.MetadataAlias{
-				Pos:   goPosToSchemaPos(pos),
+				Pos:   goPosToSchemaPos(f.Pos()),
 				Kind:  schema.AliasKindJSON,
 				Alias: jsonFieldName,
 			})
 		}
 		out.Fields = append(out.Fields, &schema.Field{
-			Pos:      goPosToSchemaPos(pos),
+			Pos:      goPosToSchemaPos(f.Pos()),
```
Comment (wesbillman) [inline go-runtime/compile/schema.go:638]: I think these all needed to be `f.Pos()` since we're validating the `s.Field(i)` here.

**Pull request merged**